website/addons/gitlab/views.py

```diff
@@ -0,0 +1,444 @@
+import httplib as http
+import logging
+
+from flask import request, make_response
+
+from framework.exceptions import HTTPError
+
+from modularodm import Q
+from modularodm.storage.base import KeyExistsException
+from website.oauth.models import ExternalAccount
+
+from website.addons.base import generic_views
+from website.addons.gitlab.api import GitLabClient, ref_to_params
+from website.addons.gitlab.exceptions import NotFoundError, GitLabError
+from website.addons.gitlab.settings import DEFAULT_HOSTS
+from website.addons.gitlab.serializer import GitLabSerializer
+from website.addons.gitlab.utils import (
+    get_refs, check_permissions,
+    verify_hook_signature, MESSAGES
+)
+
+from website.models import NodeLog
+from website.project.decorators import (
+    must_have_addon, must_be_addon_authorizer,
+    must_have_permission, must_not_be_registration,
+    must_be_contributor_or_public, must_be_valid_project,
+)
+from website.util import rubeus
+
+from framework.auth.decorators import must_be_logged_in
+from website.util import api_url_for
+
+logger = logging.getLogger(__name__)
+
+SHORT_NAME = 'gitlab'
+FULL_NAME = 'GitLab'
+
+############
+# Generics #
+############
+
+gitlab_account_list = generic_views.account_list(
+    SHORT_NAME,
+    GitLabSerializer
+)
+
+gitlab_import_auth = generic_views.import_auth(
+    SHORT_NAME,
+    GitLabSerializer
+)
+
+def _get_folders(node_addon, folder_id):
+    pass
+
+gitlab_folder_list = generic_views.folder_list(
+    SHORT_NAME,
+    FULL_NAME,
+    _get_folders
+)
+
+gitlab_get_config = generic_views.get_config(
+    SHORT_NAME,
+    GitLabSerializer
+)
+
+gitlab_deauthorize_node = generic_views.deauthorize_node(
+    SHORT_NAME
+)
+
+gitlab_root_folder = generic_views.root_folder(
+    SHORT_NAME
+)
+
+@must_be_logged_in
+def gitlab_user_config_get(auth, **kwargs):
+    """View for getting a JSON representation of the logged-in user's
+    GitLab user settings.
+    """
+
+    user_addon = auth.user.get_addon('gitlab')
+    user_has_auth = False
+    if user_addon:
+        user_has_auth = user_addon.has_auth
+
+    return {
+        'result': {
+            'userHasAuth': user_has_auth,
+            'urls': {
+                'create': api_url_for('gitlab_add_user_account'),
+                'accounts': api_url_for('gitlab_account_list'),
+            },
+            'hosts': DEFAULT_HOSTS,
+        },
+    }, http.OK
+
+@must_be_logged_in
+def gitlab_add_user_account(auth, **kwargs):
+    """Verifies new external account credentials and adds to user's list"""
+    user = auth.user
+
+    host = request.json.get('host').rstrip('/')
+    clientId = request.json.get('clientId')
+    clientSecret = request.json.get('clientSecret')
+
+    try:
+        account = ExternalAccount(
+            provider='gitlab',
+            provider_name='GitLab',
+            display_name=host,
+            oauth_key=clientId,
+            oauth_secret=clientSecret,
+            provider_id=host,
```

Comment:
This will be problematic. `ExternalAccount` objects have a uniqueness constraint over `(provider, provider_id)` -- this would prevent more than one GitLab user per GitLab host from being able to connect.